velox/expression/CastExpr.cpp

```diff
@@ -118,6 +118,32 @@ VectorPtr castFromDate(
   }
 }
 
+template <bool adjustForTimeZone>
```

Comment:
The loop body is pretty complex with conditionals and %. One more if is not going to make a big difference. I suggest to remove this template parameter.

Reply:
This conditional will be executed for each element of a row. We have seen one condition can make a difference in while executed for each row. Refer: https://github.com/facebookincubator/velox/pull/5455/files#diff-27b090936868d7e88b5053541405438a3b79ba2355bc153789cd48581836e542L50.

I have refactored code to avoid new templated function. Instead using a callback to convert to timezone. Lmk what you think?

Reply:
Can you benchmark to double check?

Reply:
`if` as constexpr
```
WARNING: Benchmark running in DEBUG mode
============================================================================
[...]benchmarks/basic/SimpleArithmetic.cpp     relative  time/iter   iters/s
============================================================================
castTimestampDate                                          83.34ms     12.00
```

`if` Runtime check
```
WARNING: Benchmark running in DEBUG mode
============================================================================
[...]benchmarks/basic/SimpleArithmetic.cpp     relative  time/iter   iters/s
============================================================================
castTimestampDate                                          85.50ms     11.70
```
Runtime check is 2.16 ms slower than constexpr check.

Reply:
Thank you for benchmarking. Did you run an existing benchmark or added a new one? If new, would you include it in this PR?

Reply:
I added a new timestamp column under input schema of SimpleArithmeticBenchmark. Included new test as follows:

```
BENCHMARK(castTimestampDate) {
  benchmark->runSmall("cast (e as date)");
  benchmark->runMedium("cast (e as date)");
  benchmark->runLarge("cast (e as date)");
}
```
I will add a new test file for benchmarking cast expr.